pallets/proof_of_existence/src/lib.rs

```diff
@@ -164,6 +181,23 @@ pub mod pallet {
                 || (values > 0 && deadline.unwrap_or_default())
         }
 
+        fn prune_old_attestations_if_needed() {
+            let attestations = Values::<T>::iter_keys()
+                .map(|(k1, _)| k1)
+                .collect::<BTreeSet<_>>();
```

Comment:
There's a big issue here: your are reading ALL proof storage in order to find all attestations index. A mitigation could be use `iter_key_prefix` that returns just an element for each attestation and decode de id by use `prefix()` call... I but that's not a solution.

Maybe you need to introduce a new storage value for the last removed id.